### PR TITLE
Add Beit3 model

### PR DESCRIPTION
**Pull request by Leon0402**
Fixes https://github.com/huggingface/transformers/issues/22178

This a rebase based on the original PR: https://github.com/huggingface/transformers/pull/22289

Potentially Missing:
- Using Scaled Dot Product Attention (SDPA)
- BeitImageProcessorFast Support
- Beit3Backbone

### CODE REVIEW

**Comment by Leon0402** [comment]
@NielsRogge Would you be willing to support me in this second attempt (https://github.com/huggingface/transformers/pull/22289) to finally bring BEIT3 support to HF?

File: src/transformers/models/beit3/modeling_beit3.py [hunk 1]
```diff
@@ -0,0 +1,1459 @@
...
+
+        pooled_output = outputs.pooler_output if return_dict else outputs[1]
+        logits = self.classifier(pooled_output)
+
+        loss = None
+        if labels is not None:
+            loss_fct = nn.KLDivLoss(reduction="batchmean")
+            log_softmax = nn.LogSoftmax(dim=-1)
+            logits = log_softmax(logits)
+            loss = loss_fct(logits, labels.contiguous())
```
Comment (Leon0402) [inline src/transformers/models/beit3/modeling_beit3.py:1095]: This part seems a bit weird. I think for VQAv2 standard loss is BceCrossEntropy and this is what Beit3 seems to use:
https://github.com/microsoft/unilm/blob/master/beit3/engine_for_finetuning.py#L201

Comment (NielsRogge) [reply src/transformers/models/beit3/modeling_beit3.py:1095]: Feel free to adjust accordingly

File: src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+
+
+def prepare_img():
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    im = Image.open(requests.get(url, stream=True).raw)
+    return im
+
+
+def get_tokenizer():
+    return XLMRobertaTokenizer.from_pretrained("https://github.com/addf400/files/releases/download/beit3/beit3.spm")
```
Comment (Leon0402) [inline src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py]: Is there any way how I can convert this directly in a XMLRobertaTokenizerFast? I know I can use the old one here and then when loading the tokenizer from the hf checkpoint it can be converted on the fly.
But I think it would be preferable to convert it directly here.

File: src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    parser.add_argument(
+        "--pytorch_dump_folder_path", default=None, type=str, help="Path to the folder to output PyTorch model."
+    )
+    parser.add_argument(
+        "--beit3_model_type",
+        default=None,
+        type=str,
+        help="Beit3 model type, it has to be one of image_classification, vqa, visual_reasoning,"
+        "image_captioning,image_text_retrieval",
+    )
```
Comment (Leon0402) [inline src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py:374]: One thing I am overall wondering is: What is with the "base pretrained" BEiT3 image? I think with BEiT3 it is common to finetune the whole model for downstream tasks like vqa or image captioning. Thus, I think it makes sense to provide the base model too and not only the finetuned versions.

But technically the base model of course also has a head. It would probably be something like Beit3ForMaskedImageModelling. Even though it is not advertised like this. So what would be best practice here? Load the standard Beit3Model without any head and ignore certain parameters? This is what I tried locally:

Modify lines 277ff:
```
    model_state_dict = ulilm_state_dict["model"]
    model_state_dict = rename_keys(model_state_dict, beit3_model_type)

    if isinstance(model, Beit3Model):
        # For Beit3Model, we need filter beit3 keys and remove them
        model_state_dict = {k.replace("beit3.", ""): v for k, v in model_state_dict.items() if k.startswith("beit3.")}

    model.load_state_dict(model_state_dict)
   ```
   Is this okay? What would be best practise?

File: src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py (hunk 2 above)
Comment (Leon0402) [reply src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py]: Also it seems that model_max_length of the tokenizer / preprocessor is not set automatically as typically with hf models. Why is that the case? Should that not be stored in the beit3.spm or do I need to set that somewhere explicitly? For instance, I couldn't find where the VILT model sets this explicitly, thus I was assuming that huggingface can get it automatically from somewhere

File: src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py (hunk 3 above)
Comment (Leon0402) [reply src/transformers/models/beit3/convert_beit3_unilm_to_pytorch.py:374]: See newest commits

File: src/transformers/models/beit3/modeling_beit3.py [hunk 4]
```diff
@@ -0,0 +1,1459 @@
...
+            return self.image(hidden_states)
+        if split_position == 0:
+            return self.text(hidden_states)
+        image_hidden, text_hidden = torch.split(
+            hidden_states,
+            [split_position, hidden_states.size(1) - split_position],
+            dim=1,
+        )
+        image_out, text_out = self.image(image_hidden), self.text(text_hidden)
+        return torch.cat([image_out, text_out], dim=1)
```
Comment (Leon0402) [inline src/transformers/models/beit3/modeling_beit3.py:147]: Would it make sense to create a parent class MultiwayLayer that takes two networks for text and image?
I think this logic is 2 or 3 three times duplicated here.

**Comment by Leon0402** [comment]
```py
class Beit3ForQuestionAnswering(Beit3PreTrainedModel):
    def __init__(self, config):
        super().__init__(config)
        embed_dim = config.hidden_size
        self.num_labels = config.num_labels
        self.beit3 = Beit3Model(config, add_pooling_layer=True)

        self.classifier = nn.Sequential(
            nn.Linear(embed_dim, embed_dim * 2),
            nn.LayerNorm(embed_dim * 2, eps=config.layer_norm_eps),
            nn.GELU(),
            nn.Linear(embed_dim * 2, config.num_labels),
        )
        self.post_init()

    def load_state_dict(self, state_dict, *args, **kwargs):
        pos_embed_checkpoint = state_dict["beit3.encoder.embed_positions.image.weight"]
        embedding_size = pos_embed_checkpoint.shape[-1]

        num_patches = self.beit3.vision_embedding.num_patches
        num_extra_tokens = self.beit3.vision_embedding.num_position_embeddings() + 2 - num_patches

        orig_size = int((pos_embed_checkpoint.shape[-2] - num_extra_tokens) ** 0.5)
        new_size = int(num_patches**0.5)

        if orig_size != new_size:
            extra_tokens = pos_embed_checkpoint[:num_extra_tokens].unsqueeze(0)
            pos_tokens = pos_embed_checkpoint[num_extra_tokens:]
            pos_tokens = pos_tokens.reshape(1, orig_size, orig_size, embedding_size).permute(0, 3, 1, 2)
            pos_tokens = torch.nn.functional.interpolate(
                pos_tokens, size=(new_size, new_size), mode="bicubic", align_corners=False
            )
            pos_tokens = pos_tokens.permute(0, 2, 3, 1).reshape(1, new_size * new_size, embedding_size)
            new_pos_embed = torch.cat((extra_tokens, pos_tokens), dim=1).squeeze(0)
            state_dict["beit3.encoder.embed_positions.image.weight"] = new_pos_embed

        return super().load_state_dict(state_dict, *args, **kwargs)
```

@NielsRogge Could you help me here. I need something along the lines of this in the code. If someone wants to process larger or smaller images, instead of reinitializing the weights, this will interpolate the new embedding weights. It is done here https://github.com/microsoft/unilm/blob/master/beit3/utils.py#L521.
I believe this should be part of the HF model and not left to the user. But I am not sure where to put it. It seems `from_pretrained` does not use `load_state_dict`. Is there any hook in HF that could be overridden, where I can put this?

File: src/transformers/models/beit3/modeling_beit3.py (hunk 4 above)
Comment (CharlesAttend) [reply src/transformers/models/beit3/modeling_beit3.py:147]: I think that would be great, yeah!  
Adding a level of abstraction here would make the code shorter while maintaining readability.

**Comment by CharlesAttend** [comment]
Hello @Leon0402,

Thank you for working on this! This implementation could also be really useful for me at work.

I’d love to help, but I’m not quite sure how to contribute effectively. I could take care of the small subclassing tasks for `Beit3MultiwayFeedForwardNetwork`, `Beit3AttentionLinearProjection`, and `Beit3LayerNorm` if you’d like (though I realize it’s not much work).

Regarding interpolation, I noticed that it's [done during the forward pass in the DeiT implementation](https://github.com/Leon0402/transformers/blob/77e45169bf0c63cffa3207fc99a93800b00aac00/src/transformers/models/deit/modeling_deit.py#L125).  
Could that also be a viable approach for us?

**Comment by Leon0402** [comment]
Hi @CharlesAttend, 

yes, feel free to implement the abstraction for the muliway logic. Maybe the best option would be to create a PR against my branch that I can merge it, once it looks good?

Regarding interpolation logic: I think doing it in the forward pass does not work too well, because huggingface will default initialise the parameters during from_pretrained and we then do not have access anymore to the old embeddings, right?

**Comment by Rocketknight1** [comment]
cc @yonigozlan @molbap maybe?

File: src/transformers/models/beit3/modeling_beit3.py [hunk 5]
```diff
@@ -0,0 +1,1447 @@
...
+            Whether or not to return the contrastive loss.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
```
Comment (molbap) [inline src/transformers/models/beit3/modeling_beit3.py]: We now have an `auto_docstring` decorator that should alleviate a portion of these

File: src/transformers/models/beit3/modeling_beit3.py
Comment (molbap) [inline src/transformers/models/beit3/modeling_beit3.py:1]: In terms of API changes, one big change is the addition of modular transformers:  https://huggingface.co/docs/transformers/v4.45.1/en/modular_transformers 
Here I'm seeing a lot of components that are very similar to existing ones, or identical (`contrastive_loss` for instance). What's # Copied from should be removed, and a `modular_beit3.py` file should be created, with the `modeling_beit3.py` file being automatically generated from it. Feel free to look at examples across the lib!

**Review by molbap — COMMENTED** [review]
I'm down to review this, left a couple guiding comments to modernize the API!

**Comment by Leon0402** [comment]
Thanks @molbap. Do you happen to have any advice on my previous comment:

> ```python
> class Beit3ForQuestionAnswering(Beit3PreTrainedModel):
>     def __init__(self, config):
>         super().__init__(config)
>         embed_dim = config.hidden_size
>         self.num_labels = config.num_labels
>         self.beit3 = Beit3Model(config, add_pooling_layer=True)
> 
>         self.classifier = nn.Sequential(
>             nn.Linear(embed_dim, embed_dim * 2),
>             nn.LayerNorm(embed_dim * 2, eps=config.layer_norm_eps),
>             nn.GELU(),
>             nn.Linear(embed_dim * 2, config.num_labels),
>         )
>         self.post_init()
> 
>     def load_state_dict(self, state_dict, *args, **kwargs):
>         pos_embed_checkpoint = state_dict["beit3.encoder.embed_positions.image.weight"]
>         embedding_size = pos_embed_checkpoint.shape[-1]
> 
>         num_patches = self.beit3.vision_embedding.num_patches
>         num_extra_tokens = self.beit3.vision_embedding.num_position_embeddings() + 2 - num_patches
> 
>         orig_size = int((pos_embed_checkpoint.shape[-2] - num_extra_tokens) ** 0.5)
>         new_size = int(num_patches**0.5)
> 
>         if orig_size != new_size:
>             extra_tokens = pos_embed_checkpoint[:num_extra_tokens].unsqueeze(0)
>             pos_tokens = pos_embed_checkpoint[num_extra_tokens:]
>             pos_tokens = pos_tokens.reshape(1, orig_size, orig_size, embedding_size).permute(0, 3, 1, 2)
>             pos_tokens = torch.nn.functional.interpolate(
>                 pos_tokens, size=(new_size, new_size), mode="bicubic", align_corners=False
>             )
>             pos_tokens = pos_tokens.permute(0, 2, 3, 1).reshape(1, new_size * new_size, embedding_size)
>             new_pos_embed = torch.cat((extra_tokens, pos_tokens), dim=1).squeeze(0)
>             state_dict["beit3.encoder.embed_positions.image.weight"] = new_pos_embed
> 
>         return super().load_state_dict(state_dict, *args, **kwargs)
> ```
> 
> @NielsRogge Could you help me here. I need something along the lines of this in the code. If someone wants to process larger or smaller images, instead of reinitializing the weights, this will interpolate the new embedding weights. It is done here https://github.com/microsoft/unilm/blob/master/beit3/utils.py#L521. I believe this should be part of the HF model and not left to the user. But I am not sure where to put it. It seems `from_pretrained` does not use `load_state_dict`. Is there any hook in HF that could be overridden, where I can put this?

**Comment by molbap** [comment]
Hey @Leon0402, for your question: right now we don't add hooks to `from_pretrained` from within a modeling file. However if you check out https://github.com/huggingface/transformers/blob/7aa888b7fa477d13153ffbfe107dfbd6c696014a/src/transformers/models/siglip2/modeling_siglip2.py#L133 you'll find a method we previously used that resizes positional embeddings. Something similar could be ok I think. 

Feel free to ping me again for another review!
